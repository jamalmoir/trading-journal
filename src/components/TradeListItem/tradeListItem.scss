.trade-list-item {
}

.trade-list-item-cell {
  padding: 0.5rem;

  &:not(:last-child) {
  }
}

.trade-list-item-edit {
  margin-left: auto;

  &:hover {
    cursor: pointer;
  }
}