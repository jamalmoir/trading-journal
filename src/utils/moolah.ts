import Big from 'big.js';


const CURRENCY_DECIMALS = {
    'AED': Big('2'),
    'AFN': Big('2'),
    'ALL': Big('2'),
    'AMD': Big('2'),
    'ANG': Big('2'),
    'AOA': Big('2'),
    'ARS': Big('2'),
    'AUD': Big('2'),
    'AWG': Big('2'),
    'AZN': Big('2'),
    'BAM': Big('2'),
    'BBD': Big('2'),
    'BDT': Big('2'),
    'BGN': Big('2'),
    'BHD': Big('3'),
    'BIF': Big('0'),
    'BMD': Big('2'),
    'BND': Big('2'),
    'BOB': Big('2'),
    'BOV': Big('2'),
    'BRL': Big('2'),
    'BSD': Big('2'),
    'BTN': Big('2'),
    'BWP': Big('2'),
    'BYN': Big('2'),
    'BZD': Big('2'),
    'CAD': Big('2'),
    'CDF': Big('2'),
    'CHE': Big('2'),
    'CHF': Big('2'),
    'CHW': Big('2'),
    'CLF': Big('4'),
    'CLP': Big('0'),
    'CNY': Big('2'),
    'COP': Big('2'),
    'COU': Big('2'),
    'CRC': Big('2'),
    'CUC': Big('2'),
    'CUP': Big('2'),
    'CVE': Big('2'),
    'CZK': Big('2'),
    'DJF': Big('0'),
    'DKK': Big('2'),
    'DOP': Big('2'),
    'DZD': Big('2'),
    'EGP': Big('2'),
    'ERN': Big('2'),
    'ETB': Big('2'),
    'EUR': Big('2'),
    'FJD': Big('2'),
    'FKP': Big('2'),
    'GBP': Big('2'),
    'GEL': Big('2'),
    'GHS': Big('2'),
    'GIP': Big('2'),
    'GMD': Big('2'),
    'GNF': Big('0'),
    'GTQ': Big('2'),
    'GYD': Big('2'),
    'HKD': Big('2'),
    'HNL': Big('2'),
    'HRK': Big('2'),
    'HTG': Big('2'),
    'HUF': Big('2'),
    'IDR': Big('2'),
    'ILS': Big('2'),
    'INR': Big('2'),
    'IQD': Big('3'),
    'IRR': Big('2'),
    'ISK': Big('0'),
    'JMD': Big('2'),
    'JOD': Big('3'),
    'JPY': Big('0'),
    'KES': Big('2'),
    'KGS': Big('2'),
    'KHR': Big('2'),
    'KMF': Big('0'),
    'KPW': Big('2'),
    'KRW': Big('0'),
    'KWD': Big('3'),
    'KYD': Big('2'),
    'KZT': Big('2'),
    'LAK': Big('2'),
    'LBP': Big('2'),
    'LKR': Big('2'),
    'LRD': Big('2'),
    'LSL': Big('2'),
    'LYD': Big('3'),
    'MAD': Big('2'),
    'MDL': Big('2'),
    'MGA': Big('2'),
    'MKD': Big('2'),
    'MMK': Big('2'),
    'MNT': Big('2'),
    'MOP': Big('2'),
    'MRU': Big('2'),
    'MUR': Big('2'),
    'MVR': Big('2'),
    'MWK': Big('2'),
    'MXN': Big('2'),
    'MXV': Big('2'),
    'MYR': Big('2'),
    'MZN': Big('2'),
    'NAD': Big('2'),
    'NGN': Big('2'),
    'NIO': Big('2'),
    'NOK': Big('2'),
    'NPR': Big('2'),
    'NZD': Big('2'),
    'OMR': Big('3'),
    'PAB': Big('2'),
    'PEN': Big('2'),
    'PGK': Big('2'),
    'PHP': Big('2'),
    'PKR': Big('2'),
    'PLN': Big('2'),
    'PYG': Big('0'),
    'QAR': Big('2'),
    'RON': Big('2'),
    'RSD': Big('2'),
    'RUB': Big('2'),
    'RWF': Big('0'),
    'SAR': Big('2'),
    'SBD': Big('2'),
    'SCR': Big('2'),
    'SDG': Big('2'),
    'SEK': Big('2'),
    'SGD': Big('2'),
    'SHP': Big('2'),
    'SLL': Big('2'),
    'SOS': Big('2'),
    'SRD': Big('2'),
    'SSP': Big('2'),
    'STN': Big('2'),
    'SVC': Big('2'),
    'SYP': Big('2'),
    'SZL': Big('2'),
    'THB': Big('2'),
    'TJS': Big('2'),
    'TMT': Big('2'),
    'TND': Big('3'),
    'TOP': Big('2'),
    'TRY': Big('2'),
    'TTD': Big('2'),
    'TWD': Big('2'),
    'TZS': Big('2'),
    'UAH': Big('2'),
    'UGX': Big('0'),
    'USD': Big('2'),
    'USN': Big('2'),
    'UYI': Big('0'),
    'UYU': Big('2'),
    'UYW': Big('4'),
    'UZS': Big('2'),
    'VES': Big('2'),
    'VND': Big('0'),
    'VUV': Big('0'),
    'WST': Big('2'),
    'XAF': Big('0'),
    'XCD': Big('2'),
    'XOF': Big('0'),
    'XPF': Big('0'),
    'YER': Big('2'),
    'ZAR': Big('2'),
    'ZMW': Big('2'),
    'ZWL': Big('2'),
}

export type CurrencySymbol = 'AED'|'AFN'|'ALL'|'AMD'|'ANG'|'AOA'|'ARS'|'AUD'|'AWG'|'AZN'|'BAM'|'BBD'|'BDT'|'BGN'|'BHD'|'BIF'|'BMD'|'BND'|'BOB'|'BOV'|'BRL'|'BSD'|'BTN'|'BWP'|'BYN'|'BZD'|'CAD'|'CDF'|'CHE'|'CHF'|'CHW'|'CLF'|'CLP'|'CNY'|'COP'|'COU'|'CRC'|'CUC'|'CUP'|'CVE'|'CZK'|'DJF'|'DKK'|'DOP'|'DZD'|'EGP'|'ERN'|'ETB'|'EUR'|'FJD'|'FKP'|'GBP'|'GEL'|'GHS'|'GIP'|'GMD'|'GNF'|'GTQ'|'GYD'|'HKD'|'HNL'|'HRK'|'HTG'|'HUF'|'IDR'|'ILS'|'INR'|'IQD'|'IRR'|'ISK'|'JMD'|'JOD'|'JPY'|'KES'|'KGS'|'KHR'|'KMF'|'KPW'|'KRW'|'KWD'|'KYD'|'KZT'|'LAK'|'LBP'|'LKR'|'LRD'|'LSL'|'LYD'|'MAD'|'MDL'|'MGA'|'MKD'|'MMK'|'MNT'|'MOP'|'MRU'|'MUR'|'MVR'|'MWK'|'MXN'|'MXV'|'MYR'|'MZN'|'NAD'|'NGN'|'NIO'|'NOK'|'NPR'|'NZD'|'OMR'|'PAB'|'PEN'|'PGK'|'PHP'|'PKR'|'PLN'|'PYG'|'QAR'|'RON'|'RSD'|'RUB'|'RWF'|'SAR'|'SBD'|'SCR'|'SDG'|'SEK'|'SGD'|'SHP'|'SLL'|'SOS'|'SRD'|'SSP'|'STN'|'SVC'|'SYP'|'SZL'|'THB'|'TJS'|'TMT'|'TND'|'TOP'|'TRY'|'TTD'|'TWD'|'TZS'|'UAH'|'UGX'|'USD'|'USN'|'UYI'|'UYU'|'UYW'|'UZS'|'VES'|'VND'|'VUV'|'WST'|'XAF'|'XCD'|'XOF'|'XPF'|'YER'|'ZAR'|'ZMW'|'ZWL';

export class Currency {
    symbol: CurrencySymbol;
    decimals: Big;

    constructor(symbol: CurrencySymbol) {
        this.symbol = symbol;
        this.decimals = CURRENCY_DECIMALS[symbol]
    }
}

export class Money {
    amount: Big;
    currency: Currency;

    constructor(amount: string | Big, currency: CurrencySymbol) {
        this.currency = new Currency(currency);
        this.amount = Big(amount);
    }

    toString = () => {
        return this.amount.toPrecision(Number(this.currency.decimals))
    }

    _arithmetic = (op: (a: Big, b?: Big) => Big, other?: Money | Big) => {
        if (other instanceof Money && this.currency.symbol != other.currency.symbol) {
            throw Error("Currencies are different");
        }

        let b = other instanceof Money ? other.amount : other

        return new Money(op(this.amount, b), this.currency.symbol);
    }

    plus = (other: Money) => this._arithmetic((a: Big, b: Big) => a.add(b), other);
    minus = (other: Money) => this._arithmetic((a: Big, b: Big) => a.sub(b), other);
    times = (other: Money | Big) => this._arithmetic((a: Big, b: Big) => a.times(b), other);
    divide = (other: Money | Big) => this._arithmetic((a: Big, b: Big) => a.div(b), other);
    pow = (other: Big) => this._arithmetic((a: Big, b: Big) => a.pow(Number(b)), other);
    sqrt = () => this._arithmetic((a: Big) => a.sqrt());
    abs = () => this._arithmetic((a: Big) => a.abs());
}