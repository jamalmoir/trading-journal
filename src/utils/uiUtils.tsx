import * as React from 'react'
import { Button, ButtonProps } from '../components/Button'
import { Heading } from '../components/Heading'
import { confirmAlert } from 'react-confirm-alert'

export const confirmDialogue = (
  title: string,
  message: string,
  buttonOneProps: ButtonProps,
  buttonTwoProps: ButtonProps
) => {
  const confirmOptions = {
    customUI: ({ onClose }: { onClose: () => void }) => {
      return (
        <div className="confirm-dialogue dark-gray grid-x grid-margin-y">
          <Heading text={title} className="cell" />
          <p className="cell">{message}</p>
          <div className="confirm-dialogue-buttons cell grid-x grid-margin-x">
            <Button
              {...buttonOneProps}
              className={buttonOneProps.className + ' cell small medium-6'}
              onClick={() => {
                buttonOneProps.onClick()
                onClose()
              }}
            />
            <Button
              {...buttonTwoProps}
              className={buttonTwoProps.className + ' cell small medium-6'}
              onClick={() => {
                buttonTwoProps.onClick()
                onClose()
              }}
            />
          </div>
        </div>
      )
    },
  }

  confirmAlert(confirmOptions)
}
