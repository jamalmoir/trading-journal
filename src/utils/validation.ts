export const validate = (val: any, rules: {[key: string]: any}, connectedValue: any) => {
  let errors: string[] = [];

  for (let rule in rules) {
    switch (rule) {
      case 'isEmail':
        if (!emailValidator(val)) { errors.push("Invalid email") }
        break;
      case 'minLength':
        if (!minLengthValidator(val, rules[rule])) { errors.push("Too short") }
        break;
      case 'maxLength':
        if (!maxLengthValidator(val, rules[rule])) { errors.push("Too Long") }
        break;
      case 'equalTo':
        if (!equalToValidator(val, connectedValue[rule])) { errors.push("Invalid email") }
        break;
      case 'notEmpty':
        if (!notEmptyValidator(val)) { errors.push("Can't be empty") }
        break;
      default:
        errors = [];
    }
  }

  return errors;
}

const isValid = (errors: string[]) => errors.length === 0;

const emailValidator = (val: string ) => {
  return /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/
	.test(val)
}

const minLengthValidator = (val: string, minLength: number) => {
	return val.length >= minLength;
}

const maxLengthValidator = (val: string, maxLength: number) => {
	return val.length <= maxLength;
}

const equalToValidator = (val: string, checkValue: string) => {
	return val === checkValue;
}

const notEmptyValidator = (val: string) => {
	return val.trim() !== '';
}
