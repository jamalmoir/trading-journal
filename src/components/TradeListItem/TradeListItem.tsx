import * as React from 'react'

import Types from 'Types'

import './tradeListItem.scss'
import { Link } from 'react-router-dom'

interface TradeListProps {
  className?: string
  trade: Types.Trade
}

export const TradeListItem = (props: TradeListProps) => {
  return (
    <tr className={props.className + ' trade-list-item'}>
      <td className={'trade-list-item-cell'}>{props.trade.instrument}</td>
      <td className={'trade-list-item-cell'}>{props.trade.strategy}</td>
      <td className={'trade-list-item-cell'}>{props.trade.kind}</td>
      <td className={'trade-list-item-cell'}>
        {props.trade.entryDate.toLocaleDateString()}
      </td>
      <td className={'trade-list-item-cell'}>
        {props.trade.entryPrice.toString()}
      </td>
      <td className={'trade-list-item-cell'}>
        {props.trade.positionSize.toString()}
      </td>
      <td className={'trade-list-item-cell'}>
        {props.trade.stopLoss.toString()}
      </td>
      <td className={'trade-list-item-cell'}>
        {props.trade.takeProfit.toString()}
      </td>
      <td className={'trade-list-item-cell'}>
        {props.trade.pl ? props.trade.exitDate.toLocaleDateString() : ''}
      </td>
      <td className={'trade-list-item-cell'}>
        {props.trade.exitPrice ? props.trade.exitPrice.toString() : ''}
      </td>
      <td className={'trade-list-item-cell'}>
        {props.trade.pl ? props.trade.pl.toString() : ''}
      </td>
      <td className={'trade-list-item-cell'}>
        <Link to={props.trade.journalId + '/trade/' + props.trade.id}>
          <i
            className="trade-list-item-edit far fa-edit"
            data-toggle="modal"
            data-target="#journalEditModal"
          ></i>
        </Link>
      </td>
    </tr>
  )
}
