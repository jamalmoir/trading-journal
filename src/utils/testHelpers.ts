import Types from 'Types';
import { Money } from './moolah';
import Big from 'big.js';


export const testTrades: () => Types.Trade[] = () => ([
    // Winner, Long
	{
    id: '1',
    journalId: '1',
    created: new Date('2019/10/30'),
    modified: new Date('2019/10/30'),
    instrument: 'GBPJPY',
    strategy: 'Strategy 1',
    kind: 'long',
    entryDate: new Date('2019/10/30'),
    entryPrice: new Big(100),
    positionSize: new Big(1000),
    stopLoss: new Big(99),
    takeProfit: new Big(102),
    exitDate: new Date('2019/10/30'),
    exitPrice: new Big(103),
    fees: new Money('150', 'JPY'),
    pl: new Money('1850', 'JPY'),
    hitTakeProfit: true,
    mfe: new Big(102),
    mae: new Big(100),
    tags: [{id: 'monday', name: 'monday'}],
    entryComment: '',
    duringComment: '',
    exitComment: '',
    flag: false,
    entryEmotion: [{id: 'calm', name: 'calm'}],
    exitEmotion: [{id: 'collected', name: 'collected'}],
    rating: 1,
    charts: [],
    },
    // Loser, Short
	{
    id: '2',
    journalId: '1',
    created: new Date('2019/10/31'),
    modified: new Date('2019/10/31'),
    instrument: 'USDJPY',
    strategy: 'Strategy 2',
    kind: 'short',
    entryDate: new Date('2019/10/31'),
    entryPrice: new Big(90),
    positionSize: new Big(500),
    stopLoss: new Big(91),
    takeProfit: new Big(88),
    exitDate: new Date('2019/10/31'),
    exitPrice: new Big(91),
    fees: new Money('150', 'JPY'),
    pl: new Money('-650', 'JPY'),
    hitTakeProfit: false,
    mfe: new Big(89),
    mae: new Big(91),
    tags: [],
    entryComment: '',
    duringComment: '',
    exitComment: '',
    flag: true,
    entryEmotion: [{id: 'rushed', name: 'rushed'}],
    exitEmotion: [{id: 'worried', name: 'worried'}],
    rating: -1,
    charts: [],
    },
    // Unfinished, Short
	{
    id: '3',
    journalId: '1',
    created: new Date('2019/10/31'),
    modified: new Date('2019/10/31'),
    instrument: 'EURJPY',
    strategy: 'Strategy 3',
    kind: 'short',
    entryDate: new Date('2019/10/31'),
    entryPrice: new Big(90),
    positionSize: new Big(500),
    stopLoss: new Big(91),
    takeProfit: new Big(88),
    exitDate: null,
    exitPrice: null,
    fees: null,
    pl: null,
    hitTakeProfit: null,
    mfe: null,
    mae: null,
    tags: [],
    entryComment: '',
    duringComment: '',
    exitComment: '',
    flag: false,
    entryEmotion: [{id: 'rushed', name: 'rushed'}],
    exitEmotion: [],
    rating: null,
    charts: [],
    },
    // Other journal, unfinished, Short
	{
    id: '4',
    journalId: '2',
    created: new Date('2019/10/31'),
    modified: new Date('2019/10/31'),
    instrument: 'EURJPY',
    strategy: 'Strategy 3',
    kind: 'short',
    entryDate: new Date('2019/10/31'),
    entryPrice: new Big(90),
    positionSize: new Big(500),
    stopLoss: new Big(91),
    takeProfit: new Big(88),
    exitDate: null,
    exitPrice: null,
    fees: null,
    pl: null,
    hitTakeProfit: null,
    mfe: null,
    mae: null,
    tags: [],
    entryComment: '',
    duringComment: '',
    exitComment: '',
    flag: false,
    entryEmotion: [{id: 'rushed', name: 'rushed'}],
    exitEmotion: [],
    rating: null,
    charts: [],
    },
    // Other user, Short
	{
    id: '5',
    journalId: '3',
    created: new Date('2019/10/31'),
    modified: new Date('2019/10/31'),
    instrument: 'EURJPY',
    strategy: 'Strategy 3',
    kind: 'short',
    entryDate: new Date('2019/10/31'),
    entryPrice: new Big(90),
    positionSize: new Big(500),
    stopLoss: new Big(91),
    takeProfit: new Big(88),
    exitDate: null,
    exitPrice: null,
    fees: null,
    pl: null,
    hitTakeProfit: null,
    mfe: null,
    mae: null,
    tags: [],
    entryComment: '',
    duringComment: '',
    exitComment: '',
    flag: false,
    entryEmotion: [{id: 'rushed', name: 'rushed'}],
    exitEmotion: [],
    rating: null,
    charts: [],
    },
])

export const testJournals: () => Types.Journal[] = () => ([
    // Live
	{
        id: '1',
        userId: '1',
        kind: 'live',
        currency: 'JPY',
        name: 'Test Live Journal',
        created: new Date('2019/10/20'),
        modified: new Date('2019/10/20'),
        tradeCount: 0,
    },
    // Demo
	{
        id: '2',
        userId: '1',
        kind: 'demo',
        currency: 'JPY',
        name: 'Test Demo Journal',
        created: new Date('2019/10/20'),
        modified: new Date('2019/10/20'),
        tradeCount: 0,
    },
    // Other user
	{
        id: '3',
        userId: '2',
        kind: 'live',
        currency: 'JPY',
        name: 'Other User Test Live Journal',
        created: new Date('2019/10/20'),
        modified: new Date('2019/10/20'),
        tradeCount: 0,
    },
]);