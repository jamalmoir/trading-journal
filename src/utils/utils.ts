import { Tag } from "react-tag-autocomplete";
import { useState } from "react";
import Types from "Types";
import { validate } from "./validation";

export const stringsToTags = (strings: string[]): Tag[] => Array.isArray(strings) ? strings.map(s => ({id:s, name: s})) : [];

export const setUpControls = (initialControls: Types.InputControls): [Types.InputControls, (key: string, value: any) => void] => {
	const [controls, setControls]: [Types.InputControls, (controls: Types.InputControls) => void] = useState(initialControls);

	const updateControls = (key: string, value: any): void => {
		let connectedValue = {};

		if (key === null) {
			setControls(value);
			return
		}

		// equalTo rule requires a connectedValue to be compared to.
		if (controls[key].validationRules.equalTo) {
			const equalControl = controls[key].validationRules.equalTo;
			const equalValue = controls[equalControl].value;

			connectedValue = {
				...connectedValue,
				equalTo: equalValue
			};
		}

		let newControls: Types.InputControls = {
			...controls,
			[key]: {
				...controls[key],
				value: value,
				errors: validate(value, controls[key].validationRules, connectedValue),
				touched: true
			}
		};

		setControls(newControls);
	}

	return [controls, updateControls]
}