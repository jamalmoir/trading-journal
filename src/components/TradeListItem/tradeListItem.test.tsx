import * as React from 'react';
import { TradeListItem } from './TradeListItem';
import { shallow } from 'enzyme';
import { testTrades } from '../../utils/testHelpers';


describe("<TradeListItem />", () => {
  it('should match snapshot', () => {
    const component = shallow(<TradeListItem trade={ testTrades()[0] } />);
    expect(component).toMatchSnapshot();
	});
});
