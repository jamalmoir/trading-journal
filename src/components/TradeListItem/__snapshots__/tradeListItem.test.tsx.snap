// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TradeListItem /> should match snapshot 1`] = `
<tr
  className="undefined trade-list-item"
>
  <td
    className="trade-list-item-cell"
  >
    GBPJPY
  </td>
  <td
    className="trade-list-item-cell"
  >
    Strategy 1
  </td>
  <td
    className="trade-list-item-cell"
  >
    long
  </td>
  <td
    className="trade-list-item-cell"
  >
    10/30/2019
  </td>
  <td
    className="trade-list-item-cell"
  >
    100
  </td>
  <td
    className="trade-list-item-cell"
  >
    1000
  </td>
  <td
    className="trade-list-item-cell"
  >
    99
  </td>
  <td
    className="trade-list-item-cell"
  >
    102
  </td>
  <td
    className="trade-list-item-cell"
  >
    10/30/2019
  </td>
  <td
    className="trade-list-item-cell"
  >
    103
  </td>
  <td
    className="trade-list-item-cell"
  >
    1850
  </td>
  <td
    className="trade-list-item-cell"
  >
    <Link
      to="1/trade/1"
    >
      <i
        className="trade-list-item-edit far fa-edit"
        data-target="#journalEditModal"
        data-toggle="modal"
      />
    </Link>
  </td>
</tr>
`;
